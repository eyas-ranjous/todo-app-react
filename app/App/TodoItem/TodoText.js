import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import classNames from 'classnames';
import styles from './styles.css';

const TodoText = ({ todo }) => {
  const dispatch = useDispatch();
  const TodoStatus = useSelector((state) => state.TodoStatus);
  const editedTodos = useSelector((state) => state.editedTodos);
  let todoInput;

  const removeEditedTodo = () => {
    dispatch({
      type: 'REMOVE_EDITED_TODO',
      todoId: todo.id
    });
  };

  const updateTodo = () => {
    if (todoInput.value) {
      todoInput.classList.remove(styles.required);
      dispatch({
        type: 'UPDATE_TODO',
        todo: {
          id: todo.id,
          text: todoInput.value
        }
      });
    }
  };

  if (!editedTodos.has(todo.id)) {
    return (
      <span
        className={todo.status === TodoStatus.DONE
          ? styles['todo-item-striked']
          : ''
        }
      >
        {todo.text}
      </span>
    );
  }

  const handleEditKeyDown = (e) => {
    if (e.keyCode === 13) {
      updateTodo();
      removeEditedTodo();
    }
  };

  return (
    <input
      className={styles['todo-item-edit-input']}
      type="text"
      onKeyDown={handleEditKeyDown}
      defaultValue={todo.text}
      ref={(el) => { todoInput = el; }}
    />
  );
};

export default TodoText;
