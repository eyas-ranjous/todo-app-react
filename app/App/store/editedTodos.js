const editedTodos = (state = new Set(), action) => {
  if (action.type === 'EDIT_TODO') {
    return new Set(state).add(action.todoId);
  }

  if (action.type === 'REMOVE_EDITED_TODO') {
    const newState = new Set(state);
    newState.delete(action.todoId);
    return newState;
  }

  return state;
};

export default editedTodos;
