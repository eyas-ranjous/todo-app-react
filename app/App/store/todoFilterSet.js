const todoFilterSet = (state = new Set(), action) => {
  if (action.type === 'UPDATE_STATUS_FILTER_SET') {
    return new Set(action.status);
  }

  return state;
};

export default todoFilterSet;
