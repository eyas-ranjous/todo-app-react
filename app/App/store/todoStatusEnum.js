export const TodoStatusEnum = {
  TODO: 1,
  DONE: 2
};

export const TodoStatus = (state = TodoStatusEnum) => state;
