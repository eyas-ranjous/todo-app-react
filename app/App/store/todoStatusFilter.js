const todoStatusFilter = (state = new Set(), action) => {
  if (action.type === 'SET_TODO_STATUS_FILTER') {
    return new Set(action.status);
  }

  return state;
};

export default todoStatusFilter;
