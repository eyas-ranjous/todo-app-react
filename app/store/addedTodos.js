import { TodoStatusEnum } from './todoStatusEnum';

let nextTodoId = 1;

const addedTodos = (state = new Map(), action) => {
  if (action.type === 'ADD_TODO') {
    const { todo } = action;
    const id = `todo-${nextTodoId}`;
    const status = TodoStatusEnum.TODO;
    nextTodoId += 1;
    return new Map(state).set(id, { ...todo, ...{ id, status } });
  }

  if (action.type === 'UPDATE_TODO') {
    const { todo } = action;

    if (!state.has(todo.id)) return state;

    const updatedTodo = { ...state.get(todo.id), ...todo };
    return new Map(state).set(todo.id, updatedTodo);
  }

  if (action.type === 'REMOVE_TODO') {
    const newState = new Map(state);
    newState.delete(action.todoId);
    return newState;
  }

  if (action.type === 'TOGGLE_TODO_STATUS') {
    if (!state.has(action.todoId)) return state;

    const todo = state.get(action.todoId);
    const { DONE, TODO } = TodoStatusEnum;
    const newStatus = todo.status === TODO ? DONE : TODO;
    const updatedTodo = { ...todo, ...{ status: newStatus } };

    return new Map(state).set(todo.id, updatedTodo);
  }

  return state;
};

export default addedTodos;
