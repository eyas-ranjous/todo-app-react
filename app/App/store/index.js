import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatus';
import todoStatusFilter from './todoStatusFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  todoStatusFilter
});

export default createStore(reducers);
