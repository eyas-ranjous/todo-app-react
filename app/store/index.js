import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatusEnum';
import { TodoFilters } from './todoFiltersEnum';
import todoFilterSet from './todoFilterSet';
import selectedFilter from './selectedFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  TodoFilters,
  todoFilterSet,
  selectedFilter
});

export default createStore(reducers);
