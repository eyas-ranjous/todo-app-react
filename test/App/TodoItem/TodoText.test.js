import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import createStore from '../../../app/store';
import { TodoStatusEnum } from '../../../app/store/TodoStatusEnum';
import TodoText from '../../../app/App/TodoItem/TodoText';

describe('TodoText', () => {
  const store = createStore();
  const todo = {
    id: 'todo-1',
    text: 'test task',
    status: TodoStatusEnum.TODO
  };

  it('renders successfully', () => {
    const wrapper = mount(
      <Provider store={store}>
        <TodoText todo={todo} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('enable editing todo text when edit clicked', () => {
    store.dispatch({ type: 'ADD_TODO', todo: { text: 'test task' } });
    store.dispatch({ type: 'EDIT_TODO', todoId: todo.id });

    const wrapper = mount(
      <Provider store={store}>
        <TodoText todo={todo} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
    const input = wrapper.find('input').getDOMNode();
    expect(input.value).toEqual(todo.text);
    input.value = 'test task updated';
    wrapper.find('input').simulate('keydown', { keyCode: 13 });
    expect(store.getState().addedTodos.get(todo.id).text)
      .toEqual('test task updated');
    expect(store.getState().editedTodos.has(todo.id)).toEqual(false);
  });
});
