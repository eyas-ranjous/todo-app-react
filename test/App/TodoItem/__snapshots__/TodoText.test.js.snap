// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoText enable editing todo text when edit clicked 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <TodoText
    todo={
      Object {
        "id": "todo-1",
        "status": 1,
        "text": "test task",
      }
    }
  >
    <input
      className="todo-item-edit-input"
      defaultValue="test task"
      onKeyDown={[Function]}
      type="text"
    />
  </TodoText>
</Provider>
`;

exports[`TodoText renders successfully 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <TodoText
    todo={
      Object {
        "id": "todo-1",
        "status": 1,
        "text": "test task",
      }
    }
  >
    <span
      className=""
    >
      test task
    </span>
  </TodoText>
</Provider>
`;
